### ISSUE TITLE
How do you identify records for update?

### ISSUE THREAD

**Issue by Kirill-H** [forcedotcom/SFDX-Data-Move-Utility#239]
I'm trying to upsert Product2 records - using an actual unique external id field as the externalId with an Upsert operation.
The operation ends with a success, but records that should be updated are not updated - and I get a file in the target library indicating that most of the records failed due to a duplication error on the externalId field....

I would expect it to update those records instead of inserting them.

This is my export file:

`{
  "objects" : [  {
    "query" : "SELECT name, Team_Product_Availability__c, productcode, description, isactive, family, displayurl, quantityunitofmeasure, stockkeepingunit, sbqq__assetamendmentbehavior__c, sbqq__assetconversion__c, sbqq__batchquantity__c, sbqq__billingfrequency__c, sbqq__billingtype__c, sbqq__blockpricingfield__c, sbqq__chargetype__c, sbqq__component__c, sbqq__compounddiscountrate__c, sbqq__configurationevent__c, sbqq__configurationfieldset__c, sbqq__configurationfields__c, sbqq__configurationformtitle__c, sbqq__configurationtype__c, sbqq__configurationvalidator__c, sbqq__configuredcodepattern__c, sbqq__configureddescriptionpattern__c, sbqq__costeditable__c, sbqq__costschedule__c, sbqq__customconfigurationpage__c, sbqq__customconfigurationrequired__c, sbqq__customercommunityavailability__c, sbqq__defaultpricingtable__c, sbqq__defaultquantity__c, sbqq__descriptionlocked__c, sbqq__discountcategory__c, sbqq__discountschedule__c, sbqq__dynamicpricingconstraint__c, sbqq__excludefrommaintenance__c, sbqq__excludefromopportunity__c, sbqq__externallyconfigurable__c, sbqq__generatecontractedprice__c, sbqq__hasconfigurationattributes__c, sbqq__hidden__c, sbqq__hidepriceinsearchresults__c, sbqq__includeinmaintenance__c, sbqq__newquotegroup__c, sbqq__nondiscountable__c, sbqq__nonpartnerdiscountable__c, sbqq__optionlayout__c, sbqq__optionselectionmethod__c, sbqq__optional__c, sbqq__priceeditable__c, sbqq__pricingmethodeditable__c, sbqq__pricingmethod__c, sbqq__productpictureid__c, sbqq__quantityeditable__c, sbqq__quantityscale__c, sbqq__reconfigurationdisabled__c, sbqq__renewalproduct__c, sbqq__sortorder__c, sbqq__specifications__c, sbqq__subscriptionbase__c, sbqq__subscriptioncategory__c, sbqq__subscriptionpercent__c, sbqq__subscriptionpricing__c, sbqq__subscriptiontarget__c, sbqq__subscriptionterm__c, sbqq__subscriptiontype__c, sbqq__taxcode__c, sbqq__taxable__c, sbqq__termdiscountlevel__c, sbqq__termdiscountschedule__c, sbqq__upgradecredit__c, sbqq__upgraderatio__c, sbqq__upgradesource__c, sbqq__upgradetarget__c, sbqq__enablelargeconfiguration__c, sbqq__hasconsumptionschedule__c, sbqq__pricingguidance__c, Team_Product_Availability__c, product_type1__c, product_discount__c, region__c, product_type__c, isspotline__c, brand__c, product_sub_type__c, ischargeline__c, product_offering__c, exclude_systemdiscounts__c, rate_type__c, deployment_key__c FROM Product2 WHERE Deploy__c = true",
    "operation" : "Upsert",
    "externalId" : "deployment_key__c"
  }
]
}`

**Comment by hknokh** [comment]
Hello @Kirill-H 
In order to analyse your case please provide:
- the complete .log file
- the dump of the target file containing the error message (at least 1-2 complete rows from this file)
- what is the exact name of the target file? For example it may be "Product2_insert_target.csv" or "Product2_update_target.csv", etc.

**Comment by Kirill-H** [comment]
> Hello @Kirill-H
> In order to analyse your case please provide:
> 
> * the complete .log file
> * the dump of the target file containing the error message (at least 1-2 complete rows from this file)
> * what is the exact name of the target file? For example it may be "Product2_insert_target.csv" or "Product2_update_target.csv", etc.

Thanks for the response @hknokh !

The target file name is Product2_insert_target.csv as you suggested.
Please see the requested files below.

[2020-11-27__15_27_08.log](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/5611693/2020-11-27__15_27_08.log)

[Product2_insert_target.txt](https://github.com/forcedotcom/SFDX-Data-Move-Utility/files/5611697/Product2_insert_target.txt)

**Comment by hknokh** [comment]
Hello, @Kirill-H 
As I understand you have problem with the letter case in the external id (deployment_key__c)
The external id field definition shoul always be in a right case, cause it is always **case sensitive** property, so try to use Deployment_Key__c instead.
You need to understand that the Plugin is not only querying the API for the records that in salesforce all queries are case insensitive, but the Tool do a lot of side jobs in transforming and analysing the queries and parameters.
The Plugin need to understand which field you want to transfer.
That is why, I would suggest you to always use right case in your queries, both for the field and for the object names.
Although the Plugin can do a name resolution for a query, sometimes it can make a mistake and not able to detect the correct field which is matching  the provided external id definition.
In summary, fix your query and external id property regarding the spelling.
